3.6.0 fails to compile with GCC 12 probably

### Checklist

- [X] I have followed every step on the [compiling wiki page](https://github.com/polybar/polybar/wiki/Compiling) and installed all necessary dependencies.
- [X] My problem is not described on the [known issues page](https://github.com/polybar/polybar/wiki/Known-Issues)
- [X] I have searched for other open and closed [issues](https://github.com/polybar/polybar/issues?q=is%3Aissue) that may have already reported this problem.
- [X] I was able to reproduce this build issue in a clean build

### From where are you building polybar?

From a release archive

### Describe how you are building polybar.

Official Fedora build

### Version

3.6.0

### Build Process

```shell
%build
%cmake
%cmake_build


%install
%cmake_install
```


### Build log

```text
[ 30%] Building CXX object bin/CMakeFiles/poly.dir/ipc/decoder.cpp.o
cd /builddir/build/BUILD/polybar-3.6.0/redhat-linux-build/bin && /usr/bin/ccache /usr/lib64/ccache/g++  -I/builddir/build/BUILD/polybar-3.6.0/include -I/builddir/build/BUILD/polybar-3.6.0/redhat-linux-build/generated-sources -isystem /usr/include/cairo -isystem /usr/include/glib-2.0 -isystem /usr/lib64/glib-2.0/include -isystem /usr/include/sysprof-4 -isystem /usr/include/pixman-1 -isystem /usr/include/freetype2 -isystem /usr/include/libpng16 -isystem /usr/include/harfbuzz -isystem /usr/include/libxml2 -isystem /builddir/build/BUILD/polybar-3.6.0/lib/xpp/include -isystem /builddir/build/BUILD/polybar-3.6.0/redhat-linux-build/lib/xpp/generated-sources/include -isystem /builddir/build/BUILD/polybar-3.6.0/lib/i3ipcpp/3rd/auss/include -isystem /builddir/build/BUILD/polybar-3.6.0/lib/i3ipcpp/include -isystem /usr/include/libnl3 -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -DNDEBUG -Wall -Wextra -Wpedantic -Wsuggest-override -std=c++14 -MD -MT bin/CMakeFiles/poly.dir/ipc/decoder.cpp.o -MF CMakeFiles/poly.dir/ipc/decoder.cpp.o.d -o CMakeFiles/poly.dir/ipc/decoder.cpp.o -c /builddir/build/BUILD/polybar-3.6.0/src/ipc/decoder.cpp
In file included from /builddir/build/BUILD/polybar-3.6.0/include/ipc/decoder.hpp:6,
                 from /builddir/build/BUILD/polybar-3.6.0/src/ipc/decoder.cpp:1:
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:18:43: error: variable 'constexpr const std::array<unsigned char, 7> polybar::ipc::MAGIC' has initializer but incomplete type
   18 |   static constexpr std::array<uint8_t, 7> MAGIC = {'p', 'o', 'l', 'y', 'i', 'p', 'c'};
      |                                           ^~~~~
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:60:32: error: static assertion failed
   60 |   static_assert(sizeof(header) == HEADER_SIZE, "");
      |                 ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:60:32: note: the comparison reduces to '(6 == 13)'
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:61:45: error: static assertion failed
   61 |   static_assert(sizeof(header::header_data) == HEADER_SIZE, "");
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:61:45: note: the comparison reduces to '(6 == 13)'
/builddir/build/BUILD/polybar-3.6.0/src/ipc/decoder.cpp: In member function 'ssize_t polybar::ipc::decoder::process_header_data(const uint8_t*, std::size_t)':
/builddir/build/BUILD/polybar-3.6.0/src/ipc/decoder.cpp:89:95: error: 'struct polybar::ipc::header::header_data' has no member named 'magic'
   89 |           "Received full ipc header (magic=%.*s version=%d size=%zd)", MAGIC.size(), header.s.magic, version, msg_size);
      |                                                                                               ^~~~~
/builddir/build/BUILD/polybar-3.6.0/src/ipc/decoder.cpp:91:27: error: 'struct polybar::ipc::header::header_data' has no member named 'magic'
   91 |       if (memcmp(header.s.magic, MAGIC.data(), MAGIC.size()) != 0) {
      |                           ^~~~~
/builddir/build/BUILD/polybar-3.6.0/src/ipc/decoder.cpp:93:67: error: 'struct polybar::ipc::header::header_data' has no member named 'magic'
   93 |                     string(reinterpret_cast<const char*>(header.s.magic), MAGIC.size()) + "'");
      |                                                                   ^~~~~
gmake[2]: *** [bin/CMakeFiles/poly.dir/build.make:387: bin/CMakeFiles/poly.dir/ipc/decoder.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
In file included from /builddir/build/BUILD/polybar-3.6.0/include/ipc/decoder.hpp:6,
                 from /builddir/build/BUILD/polybar-3.6.0/include/ipc/ipc.hpp:9,
                 from /builddir/build/BUILD/polybar-3.6.0/src/ipc/ipc.cpp:1:
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:18:43: error: variable 'constexpr const std::array<unsigned char, 7> polybar::ipc::MAGIC' has initializer but incomplete type
   18 |   static constexpr std::array<uint8_t, 7> MAGIC = {'p', 'o', 'l', 'y', 'i', 'p', 'c'};
      |                                           ^~~~~
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:60:32: error: static assertion failed
   60 |   static_assert(sizeof(header) == HEADER_SIZE, "");
      |                 ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:60:32: note: the comparison reduces to '(6 == 13)'
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:61:45: error: static assertion failed
   61 |   static_assert(sizeof(header::header_data) == HEADER_SIZE, "");
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/builddir/build/BUILD/polybar-3.6.0/include/ipc/msg.hpp:61:45: note: the comparison reduces to '(6 == 13)'
gmake[2]: *** [bin/CMakeFiles/poly.dir/build.make:373: bin/CMakeFiles/poly.dir/ipc/ipc.cpp.o] Error 1
gmake[2]: Leaving directory '/builddir/build/BUILD/polybar-3.6.0/redhat-linux-build'
gmake[1]: *** [CMakeFiles/Makefile2:523: bin/CMakeFiles/poly.dir/all] Error 2
gmake[1]: Leaving directory '/builddir/build/BUILD/polybar-3.6.0/redhat-linux-build'
gmake: *** [Makefile:139: all] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.bEDgpl (%build)
    Bad exit status from /var/tmp/rpm-tmp.bEDgpl (%build)
RPM build errors:
Child return code was: 1
EXCEPTION: [Error()]
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/mockbuild/trace_decorator.py", line 93, in trace
    result = func(*args, **kw)
  File "/usr/lib/python3.10/site-packages/mockbuild/util.py", line 599, in do_with_status
    raise exception.Error("Command failed: \n # %s\n%s" % (command, output), child.returncode)
mockbuild.exception.Error: Command failed: 
 # /usr/bin/systemd-nspawn -q -M 413e323a53d74adeafeeb76424379b9c -D /mnt/DataLinux/var/lib/mock//fedora-36-x86_64/root -a -u mockbuild --capability=cap_ipc_lock --bind=/tmp/mock-resolv.gbddcndo:/etc/resolv.conf --bind=/dev/btrfs-control --bind=/dev/loop-control --bind=/dev/loop0 --bind=/dev/loop1 --bind=/dev/loop2 --bind=/dev/loop3 --bind=/dev/loop4 --bind=/dev/loop5 --bind=/dev/loop6 --bind=/dev/loop7 --bind=/dev/loop8 --bind=/dev/loop9 --bind=/dev/loop10 --bind=/dev/loop11 --console=pipe --setenv=TERM=vt100 --setenv=SHELL=/bin/bash --setenv=HOME=/builddir --setenv=HOSTNAME=mock --setenv=PATH=/usr/bin:/bin:/usr/sbin:/sbin --setenv=PROMPT_COMMAND=printf "\033]0;<mock-chroot>\007" --setenv=PS1=<mock-chroot> \s-\v\$  --setenv=LANG=C.UTF-8 --setenv=CCACHE_DIR=/var/tmp/ccache --setenv=CCACHE_UMASK=002 --resolv-conf=off bash --login -c /usr/bin/rpmbuild -bb --target x86_64 --nodeps /builddir/build/originals/polybar.spec
```


### Linux Distribution

Fedora 36

### Additional Context

New Polybar 3.6.0 version compiles fine on Fedora 35, but fails to build on F36 and Rawhide.

GCC version: `12.0.1-0.8.fc36`

Porting to GCC 12 https://gcc.gnu.org/gcc-12/porting_to.html.

Full [build.log](https://github.com/polybar/polybar/files/8169354/build.log).

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,7 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 ### Build
-- Compiler warning in clang-13 ([`#2613`](https://github.com/polybar/polybar/pull/2613))
+- Fixed compiler warning in Clang 13 ([`#2613`](https://github.com/polybar/polybar/pull/2613))
+- Fixed compiler error in GCC 12 ([`#2616`](https://github.com/polybar/polybar/pull/2616), [`#2614`](https://github.com/polybar/polybar/issues/2614))
 
 ## [3.6.0] - 2022-03-01
 ### Breaking

--- a/include/cairo/context.hpp
+++ b/include/cairo/context.hpp
@@ -5,6 +5,7 @@
 #include <algorithm>
 #include <cmath>
 #include <deque>
+#include <iterator>
 
 #include "cairo/font.hpp"
 #include "cairo/surface.hpp"
@@ -117,9 +118,12 @@ namespace cairo {
 
     context& operator<<(const rounded_corners& c) {
       cairo_new_sub_path(m_c);
-      cairo_arc(m_c, c.x + c.w - c.radius.top_right, c.y + c.radius.top_right, c.radius.top_right, -90 * degree, 0 * degree);
-      cairo_arc(m_c, c.x + c.w - c.radius.bottom_right, c.y + c.h - c.radius.bottom_right, c.radius.bottom_right, 0 * degree, 90 * degree);
-      cairo_arc(m_c, c.x + c.radius.bottom_left, c.y + c.h - c.radius.bottom_left, c.radius.bottom_left, 90 * degree, 180 * degree);
+      cairo_arc(
+          m_c, c.x + c.w - c.radius.top_right, c.y + c.radius.top_right, c.radius.top_right, -90 * degree, 0 * degree);
+      cairo_arc(m_c, c.x + c.w - c.radius.bottom_right, c.y + c.h - c.radius.bottom_right, c.radius.bottom_right,
+          0 * degree, 90 * degree);
+      cairo_arc(m_c, c.x + c.radius.bottom_left, c.y + c.h - c.radius.bottom_left, c.radius.bottom_left, 90 * degree,
+          180 * degree);
       cairo_arc(m_c, c.x + c.radius.top_left, c.y + c.radius.top_left, c.radius.top_left, 180 * degree, 270 * degree);
       cairo_close_path(m_c);
       return *this;
@@ -142,7 +146,8 @@ namespace cairo {
           cairo_rel_line_to(m_c, 0, segment.w);
           break;
       }
-      cairo_arc_negative(m_c, segment.x, segment.y, segment.radius - segment.w, segment.angle_to * degree, segment.angle_from * degree);
+      cairo_arc_negative(m_c, segment.x, segment.y, segment.radius - segment.w, segment.angle_to * degree,
+          segment.angle_from * degree);
       cairo_close_path(m_c);
       return *this;
     }
@@ -360,9 +365,9 @@ namespace cairo {
     std::deque<pair<double, double>> m_points;
     int m_activegroups{0};
 
-    private:
-      const double degree = M_PI / 180.0;
+   private:
+    const double degree = M_PI / 180.0;
   };
-}  // namespace cairo
+} // namespace cairo
 
 POLYBAR_NS_END

--- a/include/ipc/msg.hpp
+++ b/include/ipc/msg.hpp
@@ -2,6 +2,8 @@
 
 #include "common.hpp"
 
+#include <array>
+
 POLYBAR_NS
 
 /**

--- a/include/modules/date.hpp
+++ b/include/modules/date.hpp
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <atomic>
 #include <ctime>
 #include <iomanip>
 #include <iostream>
@@ -44,6 +45,6 @@ namespace modules {
 
     std::atomic<bool> m_toggled{false};
   };
-}  // namespace modules
+} // namespace modules
 
 POLYBAR_NS_END

--- a/include/modules/fs.hpp
+++ b/include/modules/fs.hpp
@@ -1,7 +1,5 @@
 #pragma once
 
-#include <utility>
-
 #include "components/config.hpp"
 #include "modules/meta/timer_module.hpp"
 #include "settings.hpp"
@@ -74,6 +72,6 @@ namespace modules {
     // used while formatting output
     size_t m_index{0_z};
   };
-}  // namespace modules
+} // namespace modules
 
 POLYBAR_NS_END

--- a/include/modules/github.hpp
+++ b/include/modules/github.hpp
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <atomic>
+
 #include "modules/meta/timer_module.hpp"
 #include "settings.hpp"
 #include "utils/http.hpp"
@@ -37,6 +39,6 @@ namespace modules {
     bool m_empty_notifications{false};
     std::atomic<bool> m_offline{false};
   };
-}  // namespace modules
+} // namespace modules
 
 POLYBAR_NS_END

--- a/src/adapters/net_nl.cpp
+++ b/src/adapters/net_nl.cpp
@@ -1,9 +1,10 @@
-#include "adapters/net.hpp"
-
 #include <linux/nl80211.h>
 #include <netlink/genl/ctrl.h>
 #include <netlink/genl/genl.h>
 
+#include <algorithm>
+
+#include "adapters/net.hpp"
 #include "utils/file.hpp"
 
 POLYBAR_NS
@@ -166,8 +167,8 @@ namespace net {
     m_essid.clear();
 
     if (bss[NL80211_BSS_INFORMATION_ELEMENTS] != nullptr) {
-      // Information Element ID from ieee80211.h
-      #define WLAN_EID_SSID 0
+// Information Element ID from ieee80211.h
+#define WLAN_EID_SSID 0
 
       auto ies = static_cast<char*>(nla_data(bss[NL80211_BSS_INFORMATION_ELEMENTS]));
       auto ies_len = nla_len(bss[NL80211_BSS_INFORMATION_ELEMENTS]);
@@ -226,6 +227,6 @@ namespace net {
       m_signalstrength.max = hardware_max - hardware_min;
     }
   }
-}  // namespace net
+} // namespace net
 
 POLYBAR_NS_END

--- a/src/modules/fs.cpp
+++ b/src/modules/fs.cpp
@@ -3,6 +3,7 @@
 #include <sys/statvfs.h>
 
 #include <fstream>
+#include <utility>
 
 #include "drawtypes/label.hpp"
 #include "drawtypes/progressbar.hpp"
@@ -206,6 +207,6 @@ namespace modules {
 
     return true;
   }
-}  // namespace modules
+} // namespace modules
 
 POLYBAR_NS_END

